### DISCUSSION TITLE
Design principles for Research Software?

### DISCUSSION THREAD

**Discussion by cynthiahqy** [RSEAA/RSEAA.github.io#63] [Discussion topics for #RSEAA2023]
Research software often arises from code written for domain-specific problems. The motivation or problem statement for the tool is often also written in domain-specific terms limiting the cross-domain visibility/interpretability.

**Are there software design principles or frameworks that can be adapted for designing, generalising or communicating research software contributions?**

For example, this paper offers a framework for designing visualisation tools with 4 nested components:
- domain problem characterisation
- data/operation abstraction
- encoding design
- algorithm implementation
> Munzner, Tamara. (2010). A Nested Model for Visualization Design and Validation. Visualization and Computer Graphics, IEEE Transactions on. 15. 921 - 928. 10.1109/TVCG.2009.111.

**Comment by nrmay** [comment]
@cynthiahqy Yes. Software design is a well established discipline: https://en.wikipedia.org/wiki/Software_design

A useful exercise may be to identify appropriate Design Patterns (https://en.wikipedia.org/wiki/Software_design_pattern) for the research domain. 
Note: design patterns, especially the work by Gamma, et al. ('the gang of four'), are taught in most software engineering courses.
